src/librustc/sir.rs

```diff
@@ -0,0 +1,136 @@
+// LLVM backend expresses pointers to these using references, we use raw pointers so to as avoid
+// introducing lifetime parameters to the SirCx (and thus into TyCtxt and every place that uses
+// it).
+extern { pub type Value; }
+extern { pub type BasicBlock; }
+
+newtype_index! {
+    pub struct SirFuncIdx {
+        DEBUG_FORMAT = "SirFuncIdx({})"
+    }
+}
+
+// The index of a block within a function.
+// Note that these indices are not globally unique. For a globally unique block identifier, a
+// (SirFuncIdx, SirBlockIdx) pair must be used.
+newtype_index! {
+    pub struct SirBlockIdx {
+        DEBUG_FORMAT = "SirBlockIdx({})"
+    }
+}
+
+/// Sir equivalents of LLVM values.
+#[derive(Debug)]
+pub enum SirValue {
+    Func(SirFuncIdx),
+}
+
+impl SirValue {
+    pub fn func_idx(&self) -> SirFuncIdx {
+        let Self::Func(idx) = self;
+        *idx
+    }
+}
+
+pub struct SirCx {
+    /// Maps an opaque LLVM `Value` to its SIR equivalent.
+    pub llvm_values: FxHashMap<*const Value, SirValue>,
+    /// Maps an opaque LLVM `BasicBlock` to the function and block index of it's SIR equivalent.
+    pub llvm_blocks: FxHashMap<*const BasicBlock, (SirFuncIdx, SirBlockIdx)>,
+    /// Function store. Also owns the blocks
+    pub funcs: IndexVec<SirFuncIdx, ykpack::Body>,
+}
+
+impl SirCx {
+    pub fn new() -> Self {
+        Self {
+            llvm_values: Default::default(),
+            llvm_blocks: FxHashMap::default(),
+            funcs: Default::default(),
+        }
+    }
+
+    pub fn add_func(&mut self, value: *const Value, symbol_name: String) {
+        let idx = SirFuncIdx::from_usize(self.funcs.len());
+
+        self.funcs.push(ykpack::Body{
+            symbol_name,
+            blocks: Default::default(),
+            num_locals: 0,  // FIXME
+            num_args: 0,    // FIXME
+            flags: 0,       // Set later.
+        });
+        let existing = self.llvm_values.insert(value, SirValue::Func(idx));
+        debug_assert!(existing.is_none());
+    }
+
+    pub fn add_block(&mut self, func: *const Value, block: *const BasicBlock) {
+        let func_idx = self.llvm_values[&func].func_idx();
+        let sir_func = &mut self.funcs[func_idx];
+        let block_idx = SirBlockIdx::from_usize(sir_func.blocks.len());
+        sir_func.blocks.push(ykpack::BasicBlock{
+            stmts: Default::default(),
+            term: ykpack::Terminator::Unreachable, // FIXME
+        });
+        let existing = self.llvm_blocks.insert(block, (func_idx, block_idx));
+        debug_assert!(existing.is_none());
+    }
+
+    pub fn get_symbol_name(&mut self, func: *const Value) -> &String {
+        let func_idx = self.llvm_values[&func].func_idx();
+        let sir_func = &mut self.funcs[func_idx];
+        &sir_func.symbol_name
+    }
+
+    /// For hardware tracing, during codegen we insert DILabels to know where we are in the binary.
```

Comment:
This bit of code *probably* isn't required, as no reproducible build test actually uses `-C tracer` at present.

 I implemented this because I thought it was causing a test failure, but the real problem was elsewhere: we were erroneously creating and linking empty LLVM SIR modules when `-C tracer` is **not** used. (The SIR objects we link have non-deterministic paths).

This code may as well stay, as it doesn't noticeably slow us down and we might one day want to test reproducible builds with tracing enabled. I know how to fix it, just it's not top priority right now.